File: BackEnd/src/main/java/tn/encar/gestnotes/services/impl/DepartementService.java
package tn.encar.gestnotes.services.impl;

import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import jakarta.persistence.EntityNotFoundException;
import tn.encar.gestnotes.models.entities.Classe;
import tn.encar.gestnotes.models.entities.Departement;
import tn.encar.gestnotes.models.entities.Regle;
import tn.encar.gestnotes.repositories.ClasseRepository;
import tn.encar.gestnotes.repositories.DepartementRepository;
import tn.encar.gestnotes.repositories.RegleRepository;
import tn.encar.gestnotes.services.I_DepartementService;


@Service
public class DepartementService implements I_DepartementService{

	@Autowired
	private DepartementRepository departementRepository;
	
	@Autowired
	private ClasseRepository classeRepository;

	@Autowired
	private RegleRepository regleRepository;
	
	@Override
	public List<Departement> getDepartements() {		
		return (List<Departement>) departementRepository.findAll();
	}

	@Override
	public Departement getDepartementById(int id) {
		Departement departement = departementRepository.findById(id).orElseThrow(()->new IllegalStateException("Departement avec id "+id+" n'existe pas!"));
		return departement;
	}

	@Override
	public List<Departement> getDepartementByNom(String nom) {
		return departementRepository.findByNom(nom);
	}

	@Override
	public List<Classe> getClassesByIdDepartement(int id) {
		Departement departement = departementRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Department avec id " + id+" n'existe pas!"));
		return new ArrayList<>(departement.getClasses());
	}
	
	@Override
	public Regle getRegleByDepartementId(int id) {
		Departement departement = departementRepository.findById(id)
				.orElseThrow(() -> new EntityNotFoundException("Department avec id " + id+" n'existe pas!"));
		List<Regle> regles = (List<Regle>) regleRepository.findAll();
		for(Regle rgl : regles) {
			if(rgl.getDepartements().contains(departement))
				return rgl;
		}
 		return null;
	}

	@Override
	public void addNewDepartement(Departement departement) {
		if(!departementRepository.findByNom(departement.getNom()).isEmpty())
			throw new IllegalStateException("departement par le nom "+departement.getNom()+" existe deja!");
		departementRepository.save(departement);
		
	}

	@Override
	public void updateDepartement(int id, String nom) {
		Departement department = departementRepository.findById(id)
				.orElseThrow(() -> new EntityNotFoundException("Department avec id " + id+" n'existe pas!"));
        
        department.setNom(nom);
        departementRepository.save(department);
		
	}

	@Override
	public void deleteDepartementById(int id) {
		Departement department = departementRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Department avec id " + id+" n'existe pas!"));
        
        departementRepository.delete(department);
		
	}

	@Override
	public int countDepartement() {
		return (int) departementRepository.count();
	}

	@Override
	public int countClassesOfDepartementById(int id) {
		Departement department = departementRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Department avec id " + id+" n'existe pas!"));
        
        return department.getClasses().size();
	}

	@Override
	public Departement assignClasseToDepartement(int departementId, int classeId) {
		Departement departement = departementRepository.findById(departementId)
                .orElseThrow(() -> new EntityNotFoundException("Department avec id " + departementId+" n'existe pas!"));
        Classe classe = classeRepository.findById(classeId)
        		.orElseThrow(()->new EntityNotFoundException("Classe avec id " + classeId+" n'existe pas!"));
        List<Departement> departements = (List<Departement>) departementRepository.findAll();
        for(Departement dp : departements) {
        	if(dp.getClasses().contains(classe))
        		throw new IllegalStateException("classe already assigned!");
        }
        departement.getClasses().add(classe);
        return departementRepository.save(departement);
	}

	@Override
	public Departement removeClasseFromDepartement(int departementId, int classeId) {
		Departement departement = departementRepository.findById(departementId)
                .orElseThrow(() -> new EntityNotFoundException("Department avec id " + departementId+" n'existe pas!"));
        Classe classe = classeRepository.findById(classeId)
        		.orElseThrow(()->new EntityNotFoundException("Classe avec id " + classeId+" n'existe pas!"));
        if(!departement.getClasses().contains(classe)) {
        	throw new IllegalStateException("classe n'appartient pas a ce departement!");
        }
        departement.getClasses().remove(classe);
		return departementRepository.save(departement);
	}

}

File: FrontEnd/src/app/services/departement.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Departement } from '../models/Departement';
import { Classe } from '../models/Classe';
import { Regle } from '../models/Regle';

@Injectable({
  providedIn: 'root'
})
export class DepartementService {


  private url = 'http://localhost:8080/api/departements/'; 

  constructor(private http: HttpClient) { }

  getAllDepartements(): Observable<Departement[]> {
    return this.http.get<Departement[]>(this.url+'all');
  }

  getDepartementById(id: number): Observable<Departement> {
    return this.http.get<Departement>(this.url+'get/'+id);
  }

  getDepartementByNom(nom: string): Observable<Departement[]> {
    return this.http.get<Departement[]>(this.url+'get/nom/'+nom);
  }

  getClassesByIdDepartement(id: number): Observable<Classe[]> {
    return this.http.get<Classe[]>(this.url+'get-classes/'+id);
  }

  getRegleByDepartementId(id: number): Observable<Regle> {
    return this.http.get<Regle>(this.url+'get-regle/'+id);
  }

  addNewDepartement(departement: Departement): Observable<any> {
    return this.http.post(this.url+'add', departement);
  }

  updateDepartement(id: number, nom: string): Observable<any> {
    return this.http.put(this.url + 'update/' + id + '?nom=' + nom, {});
  }

  deleteDepartementById(id: number): Observable<any> {
    return this.http.delete(this.url+'delete/'+id);
  }

  countDepartement(): Observable<number> {
    return this.http.get<number>(this.url+'count');
  }

  countClassesOfDepartementById(id: number): Observable<number> {
    return this.http.get<number>(this.url+id+'/count-classes');
  }

  assignClasseToDepartement(departementId: number, classeId: number): Observable<Departement> {
    return this.http.post<Departement>(this.url+departementId+'/classe/'+classeId, {});
  }

  removeClasseFromDepartement(departementId: number, classeId: number): Observable<Departement> {
    return this.http.delete<Departement>(this.url+departementId+'/classe/'+classeId);
  }
}
